pkg/controller/elasticsearch/driver/pvc_expansion_test.go

```diff
@@ -0,0 +1,632 @@
+
+	sset2 := &appsv1.StatefulSet{ObjectMeta: metav1.ObjectMeta{Namespace: "ns", Name: "sset2", UID: "sset2-uid"}}
+	sset2Bytes, _ := json.Marshal(sset2)
+	sset2Json := string(sset2Bytes)
+
+	type args struct {
+		runtimeObjs []runtime.Object
+		es          esv1.Elasticsearch
+	}
+	tests := []struct {
+		name string
+		args
+		wantES          esv1.Elasticsearch
+		wantSsets       []appsv1.StatefulSet
+		wantPods        []corev1.Pod
+		wantRecreations int
+	}{
+		{
+			name: "no annotation: nothing to do",
+			args: args{
+				runtimeObjs: []runtime.Object{sset1, pod1},
+				es:          *es(),
+			},
+			wantES:          *es(),
+			wantPods:        []corev1.Pod{*pod1},
+			wantRecreations: 0,
+		},
+		{
+			name: "StatefulSet to delete",
+			args: args{
+				runtimeObjs: []runtime.Object{sset1, pod1}, // sset exists with the same UID
+				es:          *withAnnotation(es(), "elasticsearch.k8s.elastic.co/recreate-sset1", sset1Json),
+			},
+			wantES:          *withAnnotation(es(), "elasticsearch.k8s.elastic.co/recreate-sset1", sset1Json),
+			wantSsets:       nil,                             // deleted
+			wantPods:        []corev1.Pod{*pod1WithOwnerRef}, // owner ref set to the ES resource
+			wantRecreations: 1,
+		},
+		{
+			name: "StatefulSet to create",
+			args: args{
+				runtimeObjs: []runtime.Object{pod1}, // sset doesn't exist
+				es:          *withAnnotation(es(), "elasticsearch.k8s.elastic.co/recreate-sset1", sset1Json),
+			},
+			wantES: *withAnnotation(es(), "elasticsearch.k8s.elastic.co/recreate-sset1", sset1Json),
+			// created, no UUID due to how the fake client creates objects
+			wantSsets:       []appsv1.StatefulSet{{ObjectMeta: metav1.ObjectMeta{Namespace: "ns", Name: "sset1"}}},
+			wantPods:        []corev1.Pod{*pod1}, // unmodified
+			wantRecreations: 1,
+		},
+		{
+			name: "StatefulSet already recreated: remove the annotation",
+			args: args{
+				runtimeObjs: []runtime.Object{sset1DifferentUID, pod1WithOwnerRef}, // sset recreated
+				es:          *withAnnotation(es(), "elasticsearch.k8s.elastic.co/recreate-sset1", sset1Json),
+			},
+			wantES:          *es(),                                    // annotation removed
+			wantSsets:       []appsv1.StatefulSet{*sset1DifferentUID}, // same
+			wantPods:        []corev1.Pod{*pod1},                      // ownerRef removed
+			wantRecreations: 0,
+		},
+		{
+			name: "multiple statefulsets to handle",
+			args: args{
+				runtimeObjs: []runtime.Object{sset1, sset2, pod1},
+				es: *withAnnotation(withAnnotation(es(),
+					"elasticsearch.k8s.elastic.co/recreate-sset1", sset1Json),
+					"elasticsearch.k8s.elastic.co/recreate-sset2", sset2Json),
+			},
+			wantES: *withAnnotation(withAnnotation(es(),
+				"elasticsearch.k8s.elastic.co/recreate-sset1", sset1Json),
+				"elasticsearch.k8s.elastic.co/recreate-sset2", sset2Json),
+			wantSsets:       nil,
+			wantPods:        []corev1.Pod{*pod1WithOwnerRef}, // ownerRef removed
+			wantRecreations: 2,
+		},
```

Comment:
Maybe add a test with an annotation which does not contain a serialized sset to test the `!strings.HasPrefix(key, RecreateStatefulSetAnnotationPrefix)` condition ?
```suggestion
		},
				{
			name: "multiple statefulsets to handle",
			args: args{
				runtimeObjs: []runtime.Object{sset1, sset2, pod1},
				es: *withAnnotation(withAnnotation(withAnnotation(es(),
					"elasticsearch.k8s.elastic.co/recreate-sset1", sset1Json),
					"elasticsearch.k8s.elastic.co/recreate-sset2", sset2Json),
					"foo", "bar"),
			},
			wantES: *withAnnotation(withAnnotation(withAnnotation(es(),
				"elasticsearch.k8s.elastic.co/recreate-sset1", sset1Json),
				"elasticsearch.k8s.elastic.co/recreate-sset2", sset2Json),
				"foo", "bar"),
			wantSsets:       nil,
			wantPods:        []corev1.Pod{*pod1WithOwnerRef}, // ownerRef removed
			wantRecreations: 2,
		},
```

Reply:
Good idea.

Reply:
Actually I'll write a complete unit tests set for `ssetsToRecreate` that covers all the annotation cases.